File: jobs/2026-06-08__12-44-45/cron_job_cleanup_go__xFRgMJz/artifacts/user/cleanup-app/cleanup/cleanup.go
package cleanup

import (
	"context"
	"time"

	"encore.dev/cron"
	"encore.dev/storage/sqldb"
)

// Define the PostgreSQL database for this service.
var db = sqldb.NewDatabase("cleanup_db", sqldb.DatabaseConfig{
	Migrations: "./migrations",
})

// Define a cron job that runs every hour to clean up stale records.
// It calls the POST /cleanup endpoint.
var _ = cron.NewJob("cleanup-job", cron.JobConfig{
	Title:    "Clean up stale records",
	Every:    1 * cron.Hour,
	Endpoint: Cleanup,
})

// Record represents a single record in the database.
type Record struct {
	ID        string    `json:"id"`
	Data      string    `json:"data"`
	CreatedAt time.Time `json:"created_at"`
}

// InsertRequest is the request body for the POST /records endpoint.
type InsertRequest struct {
	ID        string    `json:"id"`
	Data      string    `json:"data"`
	CreatedAt time.Time `json:"created_at"`
}

// InsertResponse is the response for the POST /records endpoint.
type InsertResponse struct {
	Record Record `json:"record"`
}

// ListResponse is the response for the GET /records endpoint.
type ListResponse struct {
	Records []Record `json:"records"`
}

// CleanupResponse is the response for the POST /cleanup endpoint.
type CleanupResponse struct {
	Deleted int64 `json:"deleted"`
}

// InsertRecord inserts a new record into the database.
//
//encore:api public method=POST path=/records
func InsertRecord(ctx context.Context, req *InsertRequest) (*InsertResponse, error) {
	_, err := db.Exec(ctx,
		"INSERT INTO records (id, data, created_at) VALUES ($1, $2, $3)",
		req.ID, req.Data, req.CreatedAt,
	)
	if err != nil {
		return nil, err
	}
	return &InsertResponse{
		Record: Record{
			ID:        req.ID,
			Data:      req.Data,
			CreatedAt: req.CreatedAt,
		},
	}, nil
}

// ListRecords returns all records from the database.
//
//encore:api public method=GET path=/records
func ListRecords(ctx context.Context) (*ListResponse, error) {
	rows, err := db.Query(ctx, "SELECT id, data, created_at FROM records ORDER BY created_at DESC")
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var records []Record
	for rows.Next() {
		var r Record
		if err := rows.Scan(&r.ID, &r.Data, &r.CreatedAt); err != nil {
			return nil, err
		}
		records = append(records, r)
	}
	if err := rows.Err(); err != nil {
		return nil, err
	}
	if records == nil {
		records = []Record{}
	}
	return &ListResponse{Records: records}, nil
}

// Cleanup deletes records older than 24 hours from the database.
// This endpoint is called by the cron job and can also be triggered manually.
//
//encore:api public method=POST path=/cleanup
func Cleanup(ctx context.Context) (*CleanupResponse, error) {
	cutoff := time.Now().Add(-24 * time.Hour)
	tag, err := db.Exec(ctx,
		"DELETE FROM records WHERE created_at < $1",
		cutoff,
	)
	if err != nil {
		return nil, err
	}
	return &CleanupResponse{Deleted: tag.RowsAffected()}, nil
}

File: jobs/2026-06-08__12-44-45/cron_job_cleanup_ts__rSjTBr8/artifacts/user/cleanup/cleanup.ts
import { api } from "encore.dev/api";
import { SQLDatabase } from "encore.dev/storage/sqldb";
import { CronJob } from "encore.dev/cron";

// Define the database
const db = new SQLDatabase("records_db", {
  migrations: "./migrations",
});

// Interfaces
interface InsertRecordRequest {
  data: string;
  created_at?: string;
}

interface InsertRecordResponse {
  id: number;
}

interface Record {
  id: number;
  data: string;
  created_at: string;
}

interface CleanupResponse {
  deleted_count: number;
}

// POST /records - Insert a new record
export const insertRecord = api(
  { method: "POST", path: "/records", expose: true },
  async (req: InsertRecordRequest): Promise<InsertRecordResponse> => {
    const createdAt = req.created_at ? new Date(req.created_at) : new Date();
    const row = await db.queryRow<{ id: number }>`
      INSERT INTO records (data, created_at)
      VALUES (${req.data}, ${createdAt})
      RETURNING id
    `;
    return { id: row!.id };
  }
);

// GET /records - Return all records
export const getRecords = api(
  { method: "GET", path: "/records", expose: true },
  async (): Promise<{ records: Record[] }> => {
    const rows: Record[] = [];
    for await (const row of db.query<{ id: number; data: string; created_at: Date }>`
      SELECT id, data, created_at FROM records ORDER BY id
    `) {
      rows.push({
        id: row.id,
        data: row.data,
        created_at: row.created_at.toISOString(),
      });
    }
    return { records: rows };
  }
);

// POST /cleanup - Delete records older than 1 hour
export const cleanupRecords = api(
  { method: "POST", path: "/cleanup", expose: true },
  async (): Promise<CleanupResponse> => {
    const result = await db.queryRow<{ count: number }>`
      WITH deleted AS (
        DELETE FROM records
        WHERE created_at < NOW() - INTERVAL '1 hour'
        RETURNING id
      )
      SELECT COUNT(*) AS count FROM deleted
    `;
    return { deleted_count: result ? Number(result.count) : 0 };
  }
);

// Cron job - runs every 1 hour
const _cleanupJob = new CronJob("cleanup-job", {
  title: "Cleanup stale records",
  every: "1h",
  endpoint: cleanupRecords,
});
